src/main/java/nextstep/qna/service/QnAService.java

```diff
@@ -26,24 +24,8 @@ public class QnAService {
     @Transactional
     public void deleteQuestion(NsUser loginUser, long questionId) throws CannotDeleteException {
         Question question = questionRepository.findById(questionId).orElseThrow(NotFoundException::new);
-        if (!question.isOwner(loginUser)) {
-            throw new CannotDeleteException("질문을 삭제할 권한이 없습니다.");
-        }
-
-        List<Answer> answers = question.getAnswers();
-        for (Answer answer : answers) {
-            if (!answer.isOwner(loginUser)) {
-                throw new CannotDeleteException("다른 사람이 쓴 답변이 있어 삭제할 수 없습니다.");
-            }
-        }
-
-        List<DeleteHistory> deleteHistories = new ArrayList<>();
-        question.setDeleted(true);
-        deleteHistories.add(new DeleteHistory(ContentType.QUESTION, questionId, question.getWriter(), LocalDateTime.now()));
-        for (Answer answer : answers) {
-            answer.setDeleted(true);
-            deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now()));
-        }
+        question.delete(loginUser);
+        List<DeleteHistory> deleteHistories = question.deleteHistories();
```

Comment:
```suggestion
        List<DeleteHistory> deleteHistories = question.delete(loginUser);
```

도메인 로직으로 리팩토링 너무 잘 해주셨습니다. 👍 

더 나아가 삭제와 동시에 히스토리를 반환하도록 구현하면, 
히스토리를 따로 불러오는 작업을 생략할 수 있어서 코드의 실수를 줄일 수 있을 것 같습니다. 
이러면 히스토리를 불러오는 작업을 놓칠 가능성도 사라지게 되겠네요
관련해서 테스트 코드도 작성할 수 있어보입니다!

Reply:
감사합니다. 하나로 합친 코드로 변경해보았습니다.
테스트시 localDateTime이 테스트를 좀 어렵게 만드는 것 같아서 delete 파라미터로 주입받아서 delete 테스트도 QuestionTest에 추가해서 진행해보았습니다.

저의 경우, 메서드가 동작과 반환을 같이 하는 것 보다는 분리되는 것이 좋다는 내용을 과정중에 보았던 것으로 기억해서 고민후, 분리해보았는데요, 리뷰이님은 어떻게 생각하시나요??

Reply:
> 감사합니다. 하나로 합친 코드로 변경해보았습니다. 테스트시 localDateTime이 테스트를 좀 어렵게 만드는 것 같아서 delete 파라미터로 주입받아서 delete 테스트도 QuestionTest에 추가해서 진행해보았습니다. 

👍 

> 저의 경우, 메서드가 동작과 반환을 같이 하는 것 보다는 분리되는 것이 좋다는 내용을 과정중에 보았던 것으로 기억해서 고민후, 분리해보았는데요, 리뷰이님은 어떻게 생각하시나요??

상황에 따라 달라질 것 같아요. 이전 방법이 틀렸다는 의미는 아닙니다.

객체에 메시지를 보낸다고 했을 때, 객체 내부를 들여다보지 않고 객체 스스로 판단하기를 바라는 것이 맞다고 생각합니다.
따라서 Question 객체에 삭제 메시지를 보내고, 삭제 히스토리를 추가 호출해서 구하는 것이 아닌
Question 객체에 삭제 메시지를 보내면, 삭제 히스토리를 알아서 반환해주는 것이 OOP 스럽다로 생각합니다.

물론 복잡한 요구사항에서는 위와같이 구현할 수는 없을겁니다.
이번 미션에서는 서비스에 존재하는 트랜잭션 스크립트 방식의 로직을 
도메인 모델 패턴으로 리팩토링하는 경험을 해본다에 의미를 두면 좋을 것 같아요

`동작과 반환을 같이 하는 것 보다는 분리되는 것이 좋다` 이 부분은 어떤 상황인지 정확히 알려주시면
관련해서 얘기나눠봐도 좋을 것 같습니다. 😄